File: data/backtracking/permutations.py
def permute(nums: list[int]) -> list[list[int]]:
    """
    Permutations Backtracking Algorithm
    Time Complexity: O(N * N!)
    Space Complexity: O(N!)
    """
    res = []
    def backtrack(first=0):
        if first == len(nums):
            res.append(nums[:])
            return
        for i in range(first, len(nums)):
            nums[first], nums[i] = nums[i], nums[first]
            backtrack(first + 1)
            nums[first], nums[i] = nums[i], nums[first]

    backtrack()
    return res

File: data/combination/permutations.cpp
#include <vector>

using namespace std;

class SolutionPermutations {
    vector<vector<int>> res;
    void backtrack(vector<int>& nums, int first) {
        if (first == nums.size()) {
            res.push_back(nums);
            return;
        }
        for (int i = first; i < nums.size(); ++i) {
            swap(nums[first], nums[i]);
            backtrack(nums, first + 1);
            swap(nums[first], nums[i]);
        }
    }
public:
    vector<vector<int>> permute(vector<int>& nums) {
        backtrack(nums, 0);
        return res;
    }
};
